File: backend/auth-service/src/main/java/com/optica/auth/service/AuthService.java
package com.optica.auth.service;

import com.optica.auth.dto.AuthResponse;
import com.optica.auth.dto.LoginRequest;
import com.optica.auth.dto.RegisterRequest;
import com.optica.auth.dto.UserResponse;
import com.optica.auth.exception.ApiException;
import com.optica.auth.model.Role;
import com.optica.auth.model.User;
import com.optica.auth.repository.RoleRepository;
import com.optica.auth.repository.UserRepository;
import com.optica.auth.security.JwtUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;

    public UserResponse register(RegisterRequest req) {
        if (userRepository.existsByEmail(req.getEmail())) {
            throw new ApiException(HttpStatus.CONFLICT, "El email ya esta registrado");
        }

        String roleName = (req.getRole() == null || req.getRole().isBlank()) ? "USER" : req.getRole();
        Role role = roleRepository.findByName(roleName)
                .orElseThrow(() -> new ApiException(HttpStatus.BAD_REQUEST, "Rol no valido"));

        User user = User.builder()
                .name(req.getName())
                .email(req.getEmail())
                .password(passwordEncoder.encode(req.getPassword()))
                .role(role)
                .build();

        return UserResponse.from(userRepository.save(user));
    }

    public AuthResponse login(LoginRequest req) {
        User user = userRepository.findByEmail(req.getEmail())
                .orElseThrow(() -> new ApiException(HttpStatus.UNAUTHORIZED, "Credenciales invalidas"));

        if (!passwordEncoder.matches(req.getPassword(), user.getPassword())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "Credenciales invalidas");
        }

        String token = jwtUtil.generateToken(user.getId(), user.getEmail(), user.getRole().getName());
        return AuthResponse.builder()
                .token(token)
                .user(UserResponse.from(user))
                .build();
    }
}

File: backend/auth-service/src/services/authService.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const userRepository = require('../repositories/userRepository');
const roleRepository = require('../repositories/roleRepository');
const { toUserResponseDTO } = require('../dtos/authDtos');

class AuthService {
  async register({ name, email, password, role }) {
    const exists = await userRepository.findByEmail(email);
    if (exists) {
      const err = new Error('El email ya esta registrado');
      err.status = 409;
      throw err;
    }

    const roleName = role || 'USER';
    const roleEntity = await roleRepository.findByName(roleName);
    if (!roleEntity) {
      const err = new Error('Rol no valido');
      err.status = 400;
      throw err;
    }

    const hash = await bcrypt.hash(password, 10);
    const created = await userRepository.create({
      name,
      email,
      password: hash,
      role_id: roleEntity.id,
    });

    const full = await userRepository.findById(created.id);
    return toUserResponseDTO(full);
  }

  async login({ email, password }) {
    const user = await userRepository.findByEmail(email);
    if (!user) {
      const err = new Error('Credenciales invalidas');
      err.status = 401;
      throw err;
    }
    const ok = await bcrypt.compare(password, user.password);
    if (!ok) {
      const err = new Error('Credenciales invalidas');
      err.status = 401;
      throw err;
    }

    const payload = {
      id: user.id,
      email: user.email,
      role: user.role.name,
    };
    const token = jwt.sign(payload, process.env.JWT_SECRET, {
      expiresIn: process.env.JWT_EXPIRES_IN || '2h',
    });

    return { token, user: toUserResponseDTO(user) };
  }

  verifyToken(token) {
    return jwt.verify(token, process.env.JWT_SECRET);
  }
}

module.exports = new AuthService();
